File: scala-4-week/Kata.scala
object Kata {
  def binarySearch(target: Int, nums: Array[Int]): Int = {
    return getIndex(0, nums.length - 1, target, nums)
  };

  def getIndex(start: Int, end: Int, target: Int, nums: Array[Int]): Int = {
    if (start > end) return -1

    val half = (end - start) / 2 + start
    val value = nums(half)
    if (value == target) {
      return half
    } else if (value > target) {
      return getIndex(start, half - 1, target, nums)
    } else
      return getIndex(half + 1, end, target, nums)
  }

  def main(args: Array[String]): Unit = {
    val result = binarySearch(9, Array(-1, 0, 3, 5, 9, 12))
    println("expected: 4, result: " + result)
  }
}

File: clojure-6-week/src/clojure_6_week/kata.clj
(ns clojure-6-week.kata)

(defn getIndex [start end target nums]
  (println (str start " " end " " target))
  (if (> start end)
    -1
    (let [half (atom (quot (+ start end) 2))]
      (let [value (atom (get nums @half))]
        (cond
          (= target @value) @half
          (> target @value) (do
                             (swap! half inc)
                             (getIndex @half end target nums))
          (< target @value) (do
                             (swap! half dec)
                             (getIndex start @half target nums))
          )))))

(defn binarySearch [target nums]
  (getIndex 0 (dec (count nums)) target nums))

(def nums [-1 0 3 5 9 12])
(println (str "expected: 4, result: " (binarySearch 9 nums)))
